### PR TITLE
add webpack + typescript

### PR DESCRIPTION
**Pull request by jzelinskie**

### CODE REVIEW

**Comment by jmelis** [comment]
I was playing with enabling `tslint` in your branch:
https://github.com/jmelis/qontract-server/commit/c23dcb2a2e168210052425280c57e9a61fc88a60

**Comment by jmelis** [comment]
Another thing I just realised, I think we should also migrate the Dockerfile to yarn:

```diff
diff --git a/Dockerfile b/Dockerfile
index 171608f..ef395c6 100644
--- a/Dockerfile
+++ b/Dockerfile
@@ -13,7 +13,7 @@ RUN yum install centos-release-scl-rh rh-nodejs8 -y && \
 ADD . ${APP_ROOT}
 WORKDIR ${APP_ROOT}
 
-RUN npm install
+RUN yarn install
 
 EXPOSE 4000
-CMD ["npm", "start"]
+CMD ["yarn", "run", "server"]
```

By the way, come to think of it, I think we should remove the `LOAD_METHOD=fs` from `"server": "LOAD_METHOD=fs node dist/main-bundle.js schemas",`. Is there a reason why it should be there? If we keep it there, it will be hard to load from s3 for prod deployment.

File: package.json [hunk 1]
```diff
@@ -18,5 +21,14 @@
...
     "path": "^0.12.7"
+  },
+  "devDependencies": {
+    "@types/webpack": "^4.4.22",
+    "ts-loader": "^5.3.3",
+    "ts-node": "^7.0.1",
+    "tslint": "^5.12.1",
+    "typescript": "^3.2.2",
+    "webpack": "^4.28.4",
+    "webpack-cli": "^3.2.1"
```
[the remark this answers is not in the document]
Comment (jzelinskie) [reply package.json]: Are these dev tools or runtime requirements?

File: webpack.config.ts [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+let config: webpack.Configuration = {
+  entry: [
+    './server.ts',
+  ],
+  output: {
+    path: path.resolve(__dirname, 'dist'),
+    filename: '[name]-bundle.js',
+    chunkFilename: '[name]-[chunkhash].js',
+  },
+  module: {
```
[the remark this answers is not in the document]
Comment (jzelinskie) [reply webpack.config.ts]: I think I need to install nvm. I'm using node v11.3.0, which is definitely not what ships in CentOS.

File: package.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (jmelis) [reply package.json]: I got those warnings when running `yarn run dev`, so dev tools.

File: webpack.config.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (jmelis) [reply webpack.config.ts]: True... I'm actually using [node8 from softwarecollections](https://www.softwarecollections.org/en/scls/rhscl/rh-nodejs8/). I believe we could start using [node10](https://www.softwarecollections.org/en/scls/rhscl/rh-nodejs10/) (I think node10 scl was released after we started with this project).

https://github.com/app-sre/qontract-server/blob/56da1f68e544e41728fb37f0ba8aca3a961ac9dd/Dockerfile#L9-L10

File: Dockerfile [hunk 3]
```diff
@@ -1,19 +1,25 @@
...
+ENV APP_ROOT /opt/qontract-server
 ADD . ${APP_ROOT}
 WORKDIR ${APP_ROOT}
 
-RUN npm install
+RUN adduser qontract
+RUN chown -R qontract /opt/qontract-server
+USER qontract
+
+ENV PATH /opt/rh/rh-nodejs10/root/usr/bin:$PATH
```
[the remark this answers is not in the document]
Comment (jzelinskie) [reply Dockerfile]: The user has changed

[the remark this answers is not in the document]
Comment (jmelis) [reply Dockerfile]: right!

**Pull request merged**